### ISSUE TITLE
Is there a way to execute multiple hardcoded statements?

### ISSUE THREAD

**Issue by sinall** [SOCI/soci#532]
I tried session.operator<<, seems that doesn't support multiple statements query.
In my case, I want to be able to run content of a sql file which contains several statements in order to upgrade SQLite database.

Does it support? What should I do?
If it doesn't support, may I propose a new method like 'session::exec', so it can support raw sql. This is particularly valuable for DB migration. 

I notice there is a 'execude_hardcoded' function defined in sqlite3/session.cpp, I tried to declare it in my cpp file, but got 'unresolved externals' error.

**Comment by mloskot** [comment]
You may try your scenario with `core` interface http://soci.sourceforge.net/doc/3.2/interfaces.html 
but AFAIR, it is largely dependant on DBMS.

**Comment by sinall** [comment]
Yes, I checked the source code of sqlite backend, seems that only execude_hardcoded is calling 'sqlite3_exec' which is for executing raw sql (with multiple statements).

SQLite API says:

> sqlite3_exec() → A wrapper function that does sqlite3_prepare(), sqlite3_step(), sqlite3_column(), and sqlite3_finalize() for a string of one or more SQL statements.

**Comment by mloskot** [comment]
Yes, I meant not only SOCI backend but also capabilities of client libraries of individual databases and APIs SOCI use. I'm afraid you'll have to check/debug yourself if/where SOCI interferes. I don't quite remember, sorry.

**Comment by sinall** [comment]
Thanks for your reply. I checked the source code, there's no way to call existing methods to do that except downcast the session to sqlite specified session, get the connect and call sqlite API directly.

So, I propose a new method for session to provide such flexibility. For those backends who doesn't support this, throw 'UnSupporttedException'. I'm not quite familiar with other DB APIs, I guess they should have such common interface.

**Comment by vadz** [comment]
What exactly is the advantage of a new method executing all of the statements at once compared to just executing the statements one by one?

**Comment by sinall** [comment]
That's for convenient.

There's a DB migration tool called Flyway, it supports SQL based migration (https://flywaydb.org/documentation/migration/sql), I'm trying to use SOCI to do similar task.
A upgrade sql file could contain multiple statements. So there are several way to do it:
1. Break down the sql file into small sql files, one sql per statements - that will be hard to mange
2. Parse the sql and execute statements one by one - that will need more efforts
3. Depends on flyway - that means the application package needs to bundle Flyway with JRE inside

**Comment by mellery451** [comment]
related: #674

**Comment by asmwarrior** [comment]
Hi, I'm a new user of soci library. I'm using it to access a mysql database.
I also see the similar issue. My usage of the multiple statements is that I can load a `.sql` from the file, and run that `.sql` file in a batch. I mean such code:
~~~~
        std::ifstream file("create_db_table.sql", std::ios::binary);
        std::string sqlScript((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        sql << sqlScript;
~~~~
The `.sql` file is a exported from a database/table dump. So, I can reconstruct the whole table again.

**Comment by asmwarrior** [comment]
> That's for convenient.
> 
> There's a DB migration tool called Flyway, it supports SQL based migration (https://flywaydb.org/documentation/migration/sql), I'm trying to use SOCI to do similar task. A upgrade sql file could contain multiple statements. So there are several ways to do it:
> 
>     1. Break down the sql file into small sql files, one sql per statements - that will be hard to mange
> 
>     2. Parse the sql and execute statements one by one - that will need more efforts
> 
>     3. Depends on flyway - that means the application package needs to bundle Flyway with JRE inside

I think the method 2 is the way I'm going to try. But is is hard, because command normally end with semicolon `;`, but there are some other places where semicolon exits.

**Comment by vadz** [comment]
The problem is that doing this in SOCI is not trivial neither. Most DB APIs don't seem to have any equivalent to PostgreSQL `PQexec()` or SQLite3 `sqlite3_exec()` or `sqlite3_prepare_v2()` which allows to parse multiple semi-colon separated statements easily (`sqlite3_exec()` is just a trivial wrapper around it), so if we want to expose this functionality in SOCI API we need to provide some fallback for it for the other ones -- but how? Embedding a full SQL parser doesn't seem like a great idea.

**Comment by asmwarrior** [comment]
>  Embedding a full SQL parser doesn't seem like a great idea.

Hi, VZ, thanks for the reply and explanation.
Yes, I agree with this. So currently let the user does what they want(design a parser to split the SQL statements). Not the SOCI library itself.
